frame/assets-freezer/src/lib.rs

```diff
@@ -0,0 +1,361 @@
+		source: &T::AccountId,
+		dest: &T::AccountId,
+		amount: Self::Balance,
+		death: WhenDust,
+	) -> Result<Self::Balance, DispatchError> {
+		T::Assets::transfer_best_effort(asset, source, dest, amount, death)
+	}
+}
+
+impl<T: Config> fungibles::Unbalanced<T::AccountId> for Pallet<T> where
+	T::Assets: fungibles::Unbalanced<T::AccountId>
+{
+	fn set_balance(asset: Self::AssetId, who: &T::AccountId, amount: Self::Balance)
+		-> DispatchResult
+	{
+		T::Assets::set_balance(asset, who, amount)
+	}
+
+	fn set_total_issuance(asset: Self::AssetId, amount: Self::Balance) {
+		T::Assets::set_total_issuance(asset, amount)
+	}
+
+	fn decrease_balance(
+		asset: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+		keep_alive: bool,
+	) -> Result<Self::Balance, DispatchError> {
+		T::Assets::decrease_balance(asset, who, amount, keep_alive)
+	}
+
+	fn increase_balance(asset: Self::AssetId, who: &T::AccountId, amount: Self::Balance)
+		-> Result<(), DispatchError>
+	{
+		T::Assets::increase_balance(asset, who, amount)
+	}
+
+	fn decrease_balance_at_most(
+		asset: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+		keep_alive: bool,
+	) -> Self::Balance {
+		T::Assets::decrease_balance_at_most(asset, who, amount, keep_alive)
+	}
+
+	fn increase_balance_at_most(
+		asset: Self::AssetId,
+		who: &T::AccountId,
+		amount: Self::Balance,
+	) -> Self::Balance {
+		T::Assets::increase_balance_at_most(asset, who, amount)
+	}
+}
+
+impl<T: Config> fungibles::InspectHold<<T as SystemConfig>::AccountId> for Pallet<T> {
+	fn balance_on_hold(asset: AssetIdOf<T>, who: &T::AccountId) -> BalanceOf<T> {
+		T::Store::get(&(asset, who.clone())).reserved
+	}
+	fn can_hold(asset: AssetIdOf<T>, who: &T::AccountId, amount: BalanceOf<T>) -> bool {
+		// If we can withdraw without destroying the account, then we're good.
+		<Self as fungibles::Inspect<T::AccountId>>::can_withdraw(asset, who, amount) == WithdrawConsequence::Success
+	}
+	fn reducible_balance_on_hold(asset: AssetIdOf<T>, who: &T::AccountId) -> BalanceOf<T> {
+		// Figure out the most we can transfer from the balance on hold. This is basically the same
+		// as the balance on hold, but also ensures that the account actually has enough funds to
+		// be reduced and that no freezes have been placed on the asset/account.
+		let amount = T::Store::get(&(asset, who.clone())).reserved;
+		use fungibles::InspectWithoutFreezer;
+		let backing = T::Assets::reducible_balance(asset, who, true);
+		amount.min(backing)
+	}
+}
+
+impl<T: Config> fungibles::MutateHold<<T as SystemConfig>::AccountId> for Pallet<T> where
+	T::Assets: fungibles::Transfer<T::AccountId> + fungibles::InspectWithoutFreezer<T::AccountId>
+{
+	fn hold(asset: AssetIdOf<T>, who: &T::AccountId, amount: BalanceOf<T>) -> DispatchResult {
+		if !Self::can_hold(asset, who, amount) {
+			Err(TokenError::NoFunds)?
+		}
+		T::Store::mutate(
+			&(asset, who.clone()),
+			|extra| extra.reserved = extra.reserved.saturating_add(amount),
```

Comment:
Why not checked add here?

Reply:
the point is we could fail here gracefully and without worry, so might as well do so.